File: Restaurant-information-system/src/main/java/com/galaxy/Restaurantinformationsystem/Service/MenuService.java
package com.galaxy.Restaurantinformationsystem.Service;

import com.galaxy.Restaurantinformationsystem.DTO.MenuDTO;
import com.galaxy.Restaurantinformationsystem.Entity.MenuEntity;
import com.galaxy.Restaurantinformationsystem.Repository.MenuRepository;
import com.galaxy.Restaurantinformationsystem.Repository.StoreRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;

@Service
public class MenuService {

    MenuRepository menuRepository;
    StoreRepository storeRepository;

    @Autowired
    public MenuService(MenuRepository MenuRepository, StoreRepository storeRepository) {
        this.menuRepository = MenuRepository;
        this.storeRepository = storeRepository;
    }

    public MenuDTO createMenuDTO(MenuDTO MenuDTO) {
        MenuEntity MenuEntity = toEntity(MenuDTO);
        return toDTO(menuRepository.save(MenuEntity));
    }

    public MenuDTO readMenuDTO(MenuDTO MenuDTO) {
        return null;
    }

    public ArrayList<MenuDTO> readMenuInStore(Long spk) {
        ArrayList<MenuDTO> result = new ArrayList<>();
        ArrayList<MenuEntity> found = menuRepository.findByStore_SPK(spk);
        if (found != null) {
            for (MenuEntity menuEntity : found) {
                result.add(toDTO(menuEntity));
            }
        }
        return result;
    }

    public MenuDTO updateMenuDTO(MenuDTO menuDTO) {
        MenuEntity menuEntity = menuRepository.getReferenceById(menuDTO.getMPK());
        menuEntity.setName(menuDTO.getName());
        menuEntity.setStore(storeRepository.findById(menuDTO.getSPK()).get());
        menuEntity.setImage(menuEntity.getImage());
        menuEntity.setPrice(menuEntity.getPrice());

        return toDTO(menuRepository.save(menuEntity));
    }

    @Transactional
    public void deleteMenu(MenuDTO MenuDTO) {
        MenuEntity MenuEntity = toEntity(MenuDTO);
        menuRepository.delete(MenuEntity);
    }

    public MenuEntity toEntity(MenuDTO menuDTO) {
        return MenuEntity.builder()
                .MPK(menuDTO.getMPK())
                .name(menuDTO.getName())
                .price(menuDTO.getPrice())
                .image(menuDTO.getImage())
                .store(storeRepository.findById(menuDTO.getSPK()).get())
                .build();
    }

    public MenuDTO toDTO(MenuEntity menuEntity) {
        return MenuDTO.builder()
                .MPK(menuEntity.getMPK())
                .name(menuEntity.getName())
                .price(menuEntity.getPrice())
                .image(menuEntity.getImage())
                .SPK(menuEntity.getStore().getSPK())
                .build();
    }
}

File: src/main/java/com/galaxy/Restaurantinformationsystem/service/MenuService.kt
package com.galaxy.Restaurantinformationsystem.service

import com.galaxy.Restaurantinformationsystem.dto.MenuRequestDto
import com.galaxy.Restaurantinformationsystem.dto.MenuResponseDto
import com.galaxy.Restaurantinformationsystem.entity.MenuEntity
import com.galaxy.Restaurantinformationsystem.mapper.StoreMapper
import com.galaxy.Restaurantinformationsystem.repository.MenuRepository
import com.galaxy.Restaurantinformationsystem.repository.StoreRepository
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service

/**
 * 가게의 메뉴 서비스
 * @property storeRepository 가게 Repository
 * @property menuRepository 메뉴 Repository
 * @property storeMapper 가게 매퍼
 */
@Service
class MenuService(
    val storeRepository: StoreRepository,
    val menuRepository: MenuRepository,
    val storeMapper: StoreMapper,
) {
    /**
     * 가게의 메뉴 목록을 조회합니다.
     * @param storeId 가게 id
     * @return 가게의 메뉴 목록
     */
    fun getMenuDTO(storeId: Long): List<MenuResponseDto> {
        val menus: List<MenuEntity> = menuRepository.findByStoreId(storeId)
        return menus.map { menuEntity ->
            storeMapper.toResponseDto(menuEntity)
        }.toList()
    }

    /**
     * 가게의 메뉴를 등록합니다.
     * @param storeId 가게 id
     * @param menuRequestDto 메뉴 등록 요청 DTO
     */
    fun createMenuDTO(storeId: Long, menuRequestDto: MenuRequestDto): MenuResponseDto {
        val menu = storeMapper.toEntity(menuRequestDto)
        menu.store = storeRepository.findById(storeId).get()
        return storeMapper.toResponseDto(menuRepository.save(menu))
    }

    /**
     * 가게의 메뉴를 수정합니다.
     * @param storeId 가게 id
     * @param menuRequestDto 메뉴 수정 요청 DTO
     * @param menuId 메뉴 id
     * @return 수정된 메뉴
     */
    fun updateMenuDTO(storeId: Long, menuRequestDto: MenuRequestDto, menuId: Long): MenuResponseDto {
        val menu = storeMapper.toEntity(menuRequestDto)
        menu.store = storeRepository.findById(storeId).get()
        menu.id = menuId
        return storeMapper.toResponseDto(menuRepository.save(menu))
    }

    /**
     * 가게의 메뉴를 삭제합니다.
     * @param menuId 메뉴 id
     */
    @Transactional
    fun deleteMenu(menuId: Long) {
        menuRepository.deleteById(menuId)
    }
}
